File: src/lifely/normalize.py
"""Normalize raw Google Calendar API events into clean dataclasses.

Handles timezone conversion, duration calculation, and attendee processing.
"""

from datetime import datetime
from zoneinfo import ZoneInfo

from dateutil import parser as dateutil_parser

from .models import NormalizedAttendee, NormalizedEvent

DEFAULT_TIMEZONE = "America/New_York"


def normalize_events(
    raw_events: list[dict],
    user_email: str,
    timezone: str = DEFAULT_TIMEZONE,
) -> list[NormalizedEvent]:
    """Convert raw Calendar API events to normalized dataclasses.

    Args:
        raw_events: List of event dicts from Calendar API.
        user_email: The authenticated user's email (for self-detection).
        timezone: Target timezone for all times (default: America/New_York).

    Returns:
        List of NormalizedEvent objects.
    """
    tz = ZoneInfo(timezone)
    user_email_lower = user_email.lower()
    normalized = []

    for raw in raw_events:
        event = _normalize_single_event(raw, user_email_lower, tz)
        if event:
            normalized.append(event)

    return normalized


def _normalize_single_event(
    raw: dict,
    user_email_lower: str,
    tz: ZoneInfo,
) -> NormalizedEvent | None:
    """Normalize a single raw event.

    Returns None if the event should be skipped (e.g., cancelled).
    """
    # Skip cancelled events
    if raw.get("status") == "cancelled":
        return None

    # Parse start/end times
    start, all_day = _parse_event_time(raw.get("start", {}), tz)
    end, _ = _parse_event_time(raw.get("end", {}), tz)

    if not start or not end:
        return None

    # Calculate duration
    duration_minutes = (end - start).total_seconds() / 60

    # Normalize attendees
    attendees = _normalize_attendees(raw.get("attendees", []), user_email_lower)

    # Parse metadata timestamps
    created = _parse_timestamp(raw.get("created"))
    updated = _parse_timestamp(raw.get("updated"))

    return NormalizedEvent(
        id=raw["id"],
        summary=raw.get("summary"),
        description=raw.get("description"),
        start=start,
        end=end,
        all_day=all_day,
        duration_minutes=duration_minutes,
        attendees=attendees,
        organizer_email=raw.get("organizer", {}).get("email"),
        location_raw=raw.get("location"),
        created=created,
        updated=updated,
        recurring_event_id=raw.get("recurringEventId"),
    )


def _parse_event_time(time_obj: dict, tz: ZoneInfo) -> tuple[datetime | None, bool]:
    """Parse a Calendar API start/end time object.

    Args:
        time_obj: Dict with either 'dateTime' or 'date' key.
        tz: Target timezone.

    Returns:
        Tuple of (datetime in target tz, is_all_day).
    """
    if not time_obj:
        return None, False

    # All-day events use 'date' (YYYY-MM-DD)
    if "date" in time_obj:
        date_str = time_obj["date"]
        dt = datetime.strptime(date_str, "%Y-%m-%d")
        # All-day events: use midnight in target timezone
        return dt.replace(tzinfo=tz), True

    # Timed events use 'dateTime'
    if "dateTime" in time_obj:
        dt = dateutil_parser.isoparse(time_obj["dateTime"])
        # Convert to target timezone
        return dt.astimezone(tz), False

    return None, False


def _normalize_attendees(
    raw_attendees: list[dict],
    user_email_lower: str,
) -> list[NormalizedAttendee]:
    """Normalize the attendees list.

    Args:
        raw_attendees: List of attendee dicts from API.
        user_email_lower: User's email (lowercased) for self-detection.

    Returns:
        List of NormalizedAttendee objects.
    """
    attendees = []

    for raw in raw_attendees:
        email = raw.get("email", "").lower()

        # Detect self: use API's self flag, or fall back to email comparison
        is_self = raw.get("self", False) or email == user_email_lower

        attendee = NormalizedAttendee(
            email=email,
            display_name=raw.get("displayName"),
            is_self=is_self,
            response_status=raw.get("responseStatus"),
        )
        attendees.append(attendee)

    return attendees


def _parse_timestamp(timestamp_str: str | None) -> datetime | None:
    """Parse an ISO timestamp string to datetime."""
    if not timestamp_str:
        return None
    try:
        return dateutil_parser.isoparse(timestamp_str)
    except (ValueError, TypeError):
        return None

File: lifely-web-react/src/lib/stats/normalize.ts
/**
 * Normalize raw Google Calendar API events into clean typed objects.
 *
 * Handles timezone conversion, duration calculation, and attendee processing.
 * Port of Python lifely/normalize.py to TypeScript for client-side use.
 */

import type { NormalizedAttendee, NormalizedEvent } from './types';

// Raw types from Google Calendar API
export interface RawCalendarEvent {
  id: string;
  status?: string;
  summary?: string;
  description?: string;
  start?: {
    dateTime?: string;
    date?: string;
    timeZone?: string;
  };
  end?: {
    dateTime?: string;
    date?: string;
    timeZone?: string;
  };
  attendees?: Array<{
    email?: string;
    displayName?: string;
    self?: boolean;
    responseStatus?: string;
  }>;
  organizer?: {
    email?: string;
  };
  location?: string;
  created?: string;
  updated?: string;
  recurringEventId?: string;
}

/**
 * Convert raw Calendar API events to normalized objects.
 *
 * @param rawEvents - List of event dicts from Calendar API
 * @param userEmail - The authenticated user's email (for self-detection)
 * @returns List of NormalizedEvent objects
 */
export function normalizeEvents(
  rawEvents: RawCalendarEvent[],
  userEmail: string
): NormalizedEvent[] {
  const userEmailLower = userEmail.toLowerCase();
  const normalized: NormalizedEvent[] = [];

  for (const raw of rawEvents) {
    const event = normalizeSingleEvent(raw, userEmailLower);
    if (event) {
      normalized.push(event);
    }
  }

  return normalized;
}

/**
 * Normalize a single raw event.
 * Returns null if the event should be skipped (e.g., cancelled).
 */
function normalizeSingleEvent(
  raw: RawCalendarEvent,
  userEmailLower: string
): NormalizedEvent | null {
  // Skip cancelled events
  if (raw.status === 'cancelled') {
    return null;
  }

  // Parse start/end times
  const [start, allDay] = parseEventTime(raw.start);
  const [end] = parseEventTime(raw.end);

  if (!start || !end) {
    return null;
  }

  // Calculate duration in minutes
  const durationMinutes = (end.getTime() - start.getTime()) / (1000 * 60);

  // Normalize attendees
  const attendees = normalizeAttendees(raw.attendees || [], userEmailLower);

  // Parse metadata timestamps
  const created = parseTimestamp(raw.created);
  const updated = parseTimestamp(raw.updated);

  return {
    id: raw.id,
    summary: raw.summary || null,
    description: raw.description || null,
    start,
    end,
    allDay,
    durationMinutes,
    attendees,
    organizerEmail: raw.organizer?.email || null,
    locationRaw: raw.location || null,
    created,
    updated,
    recurringEventId: raw.recurringEventId || null,
  };
}

/**
 * Parse a Calendar API start/end time object.
 *
 * @returns Tuple of [Date | null, isAllDay]
 */
function parseEventTime(
  timeObj?: RawCalendarEvent['start']
): [Date | null, boolean] {
  if (!timeObj) {
    return [null, false];
  }

  // All-day events use 'date' (YYYY-MM-DD)
  if (timeObj.date) {
    const date = new Date(timeObj.date + 'T00:00:00');
    return [date, true];
  }

  // Timed events use 'dateTime'
  if (timeObj.dateTime) {
    const date = new Date(timeObj.dateTime);
    return [date, false];
  }

  return [null, false];
}

/**
 * Normalize the attendees list.
 *
 * @param rawAttendees - List of attendee objects from API
 * @param userEmailLower - User's email (lowercased) for self-detection
 * @returns List of NormalizedAttendee objects
 */
function normalizeAttendees(
  rawAttendees: NonNullable<RawCalendarEvent['attendees']>,
  userEmailLower: string
): NormalizedAttendee[] {
  const attendees: NormalizedAttendee[] = [];

  for (const raw of rawAttendees) {
    const email = (raw.email || '').toLowerCase();

    // Detect self: use API's self flag, or fall back to email comparison
    const isSelf = raw.self === true || email === userEmailLower;

    attendees.push({
      email,
      displayName: raw.displayName || null,
      isSelf,
      responseStatus: raw.responseStatus || null,
    });
  }

  return attendees;
}

/**
 * Parse an ISO timestamp string to Date.
 */
function parseTimestamp(timestampStr?: string): Date | null {
  if (!timestampStr) {
    return null;
  }
  try {
    const date = new Date(timestampStr);
    return isNaN(date.getTime()) ? null : date;
  } catch {
    return null;
  }
}

/**
 * Format a date to YYYY-MM-DD string.
 */
export function formatDate(date: Date): string {
  return date.toISOString().split('T')[0];
}

/**
 * Get weekday name from date.
 */
export function getWeekdayName(date: Date): string {
  const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  return days[date.getDay()];
}

/**
 * Get month number (1-12) from date.
 */
export function getMonth(date: Date): number {
  return date.getMonth() + 1;
}
